### PR TITLE
Add gamma functions

### PR DESCRIPTION
**Pull request by lucafedeli88**
This PR adds functions to calculate Lorentz factors (for electrons and positrons) and normalized energies (for photons).
These new functions are now used inside the functions to calculate chi parameters and to generate pairs or photons. A new test has been added and the new functions are also exposed in the python bindings.

This PR is motivated by the fact that pair production function does not test if the normalized photon energy is > 2 . Therefore, a user must have access to a method giving exactly the same normalized energy that is calculated inside this function in order to perform this test before calling the pair production function.

### CODE REVIEW

File: multi_physics/QED/QED_tests/test_picsar_gamma_functions.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    const auto p = vec3<double>{px,py,pz};
+
+    test_gamma_photons<double, unit_system::SI>(p, gamma_exp);
+    test_gamma_photons<double, unit_system::norm_lambda>(p, gamma_exp, lambda);
+    test_gamma_photons<double, unit_system::norm_omega>(p, gamma_exp, omega);
+    test_gamma_photons<double, unit_system::heaviside_lorentz>(p, gamma_exp, 1.0);
+    test_gamma_photons<float, unit_system::SI>(p, gamma_exp);
+    test_gamma_photons<float, unit_system::norm_lambda>(p, gamma_exp, lambda);
+    test_gamma_photons<float, unit_system::norm_omega>(p, gamma_exp, omega);
+    test_gamma_photons<float, unit_system::heaviside_lorentz>(p, gamma_exp, 1.0);
```
Comment (NeilZaim) [inline multi_physics/QED/QED_tests/test_picsar_gamma_functions.cpp]: Maybe you could add a function like `test_gamma_photons_all_precisions_all_units` or similar that does these 8 tests? (as you prefer, it's really minor)

File: multi_physics/QED/python_bindings/demo_python_bindings.ipynb [hunk 2]
```diff
@@ -52,17 +52,51 @@
...
+   ],
+   "source": [
+    "#Gamma functions\n",
+    "\n",
+    "gamma_how_may = 10\n",
+    "gamma_gamma = 10\n",
+    "\n",
+    "gamma_px = (np.random.rand(gamma_how_may)-0.5)*2*gamma_gamma*m_e*c;\n",
+    "gamma_py = (np.random.rand(gamma_how_may)-0.5)*2*gamma_gamma*m_e*c;\n",
+    "gamma_pz = (np.random.rand(gamma_how_may)-0.5)*2*gamma_gamma*m_e*c;\n",
```
Comment (NeilZaim) [inline multi_physics/QED/python_bindings/demo_python_bindings.ipynb]:
```suggestion
    "gamma_how_many = 10\n",
    "gamma_gamma = 10\n",
    "\n",
    "gamma_px = (np.random.rand(gamma_how_many)-0.5)*2*gamma_gamma*m_e*c;\n",
    "gamma_py = (np.random.rand(gamma_how_many)-0.5)*2*gamma_gamma*m_e*c;\n",
    "gamma_pz = (np.random.rand(gamma_how_many)-0.5)*2*gamma_gamma*m_e*c;\n",
```

**Review by NeilZaim — COMMENTED** [review]
Thanks for this PR, it looks good to me.

Just so that I understand correctly, what currently happens if the pair production function is called with a photon energy < 2 m_e c? Quickly looking at the function, it looks like we would have `gamma_ele` < 1 and eventually the square root of a negative number to compute the electron momentum, am I correct?

Is there a reason to not check `gamma_photon > 2` in `generate_breit_wheeler_pairs`?

**Comment by lucafedeli88** [comment]
> Just so that I understand correctly, what currently happens if the pair production function is called with a photon energy < 2 m_e c? Quickly looking at the function, it looks like we would have `gamma_ele` < 1 and eventually the square root of a negative number to compute the electron momentum, am I correct?
> 
> Is there a reason to not check `gamma_photon > 2` in `generate_breit_wheeler_pairs`?

Yes, that's correct: if `gamma_photon < 2` the function returns particle momenta filled with `NaN`.

I think that I did this because in WarpX if you end up calling this function you already know that `gamma_phot > 2` and I was afraid of introducing an additional branch point in the kernel. Actually, checking for `gamma_phot >= 2` won't be a problem as far as performance is of concern. However, I don't know what the function should do in this case: we cannot throw an exception in a function which could end up in a GPU kernel. The only thing that we could consider is to set the value of a `success` parameter passed by reference by the user. 
If we are really concerned about performances (but I don't think that we should be in this case), we could perform this check or not depending on a template parameter.

File: multi_physics/QED/python_bindings/demo_python_bindings.ipynb (hunk 2 above)
Comment (lucafedeli88) [reply multi_physics/QED/python_bindings/demo_python_bindings.ipynb]: `how_may` ? :laughing:  . Thanks for catching this!

File: multi_physics/QED/QED_tests/test_picsar_gamma_functions.cpp (hunk 1 above)
Comment (lucafedeli88) [reply multi_physics/QED/QED_tests/test_picsar_gamma_functions.cpp]: Ah yes, that's a good idea!

Comment (lucafedeli88) [reply multi_physics/QED/QED_tests/test_picsar_gamma_functions.cpp]: It should be OK now.

**Review by NeilZaim — APPROVED** [review]

**Comment by NeilZaim** [comment]
Regarding the behaviour when `gamma_photon < 2` I feel like it's best to abort the run as early as possible. Since we end up with `NaNs` anyways, this should simplify problem finding should this occur.

If using a success parameter is the simplest way to go, that can work (there's no way to use assertions in device code?). To me, this is definitely not urgent and can be done in a follow-up PR.

**Pull request merged**